File: pair/solution1.clj
(defn- by-pair
  [numbers sum n]
  (let [m (- sum n)
        mcnt (get numbers m)]
    (when (or (and mcnt (= m n) (> mcnt 1))
              (and mcnt (not= m n) (> mcnt 0)))
      [n m])))

(defn find-pair
  [data sum]
  (let [numbers (frequencies data)]
    (->> data
         (keep (partial by-pair numbers sum))
         first)))

;;
;; TESTS
;;

(let [data [5, 8, 1, 5, 6, 3, 4]
      tests [[10 [5 5]]
             [11 [5 6]]
             [100 nil]]]
  (doseq [[sum res] tests
          :let [res* (find-pair data sum)]
          :when (not= res* res)]
    (throw (Exception. (format "Expected %s got %s" res res*)))))

File: find-pair/solution1.py
#!/bin/python3

def find_pair(data, sum):
  numbers = {}
  
  for n in data:
    n_ = numbers.get(n, 0)
    numbers[n] = n_ + 1

  for n in data:
    m = sum - n
    m_ = numbers.get(m)

    if m_ and m == n and m_ > 1:
      return (n, m)
    elif m_ and m != n and m_ > 0:
      return (n, m)

# 
# TESTS
# 

data = [5, 8, 1, 5, 6, 3, 4]

test_suit = [[10, (5, 5)],
             [11, (5, 6)],
             [100, None]]

for (sum, res) in test_suit:
  res_ = find_pair(data, sum)
  if res_ != res:
    raise(Exception("Expected %s, got %s" % (res, res_)))